C035 - Add support for cart PS and infinite scroll; add table view to cart

dstopp opened the pull request:
c035 - add support for PS and infinite scroll to cart; added the table view whilst i was at it.

dstopp committed:
fixes to cart for perfect scrollbar, infinite scroll

dstopp committed:
no message

dstopp committed:
added nav for table on cart page

dstopp committed:
fixed container height to account for the footer height

dstopp committed:
init table on cart change

dstopp committed:
whitespace fixup

dstopp committed:
Merge branch 'new_ui' of https://github.com/SETI/pds-opus into c035_cart

dstopp committed:
updated to allow event handlers to function on both browse and cart tabs; remove id dataTable; replace with class to be more generic

dstopp committed:
more abstraction away from the specific tab to make both browse and cart use same code/function names.

dstopp committed:
Merge branch 'new_ui' of https://github.com/SETI/pds-opus into c035_cart

# Conflicts:
#	opus/application/static_media/js/browse.js

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -663,13 +660,13 @@ var o_browse = {
         let tab = `#${opus.prefs.view}`;
         let contentsView = o_browse.getScrollContainerClass();
-        let selector = (opus.prefs.browse === "dataTable") ? `#${opus.prefs.view} #dataTable tbody tr` : `#${opus.prefs.view} .gallery .thumbnail-container`;
+        let selector = (opus.prefs[browse] === "gallery") ?  `#${opus.prefs.view} .gallery .thumbnail-container` : `#${opus.prefs.view} .op-data-table tbody tr`;
```
I don't see how this can work. Isn't `browse` an undefined variable here? Maybe this is why the slider isn't working some of the time.

rfrenchseti commented on opus/application/apps/ui/templates/base.html:
```diff
@@ -101,7 +101,7 @@
                                 </li>
                                 <li class="nav-item">
-                                    <a href="#" class="metadataModal nav-link" data-toggle="modal" data-target="#metadataSelector" title="Select metadata to display or include in CSV files for each observation"><i class="fas fa-columns"></i>&nbsp;Select Metadata</a>
+                                    <a href="#" class="op-metadataModal nav-link" data-toggle="modal" data-target="#metadataSelector" title="Select metadata to display or include in CSV files for each observation"><i class="fas fa-columns"></i>&nbsp;Select Metadata</a>
```
I think we should be calling this op-metadata-modal.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1342,7 +1339,8 @@ var o_browse = {
     // return the infiniteScroll container class for either gallery or table view
     getScrollContainerClass: function() {
-        return (opus.prefs.browse === "gallery" ? ".op-gallery-view" : ".op-dataTable-view");
+        let browse = (opus.prefs.view === "cart_" ? "cart_browse" : "browse");
```
This should not have a "_" after cart

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -978,28 +976,33 @@ var o_browse = {
 
     updateBrowseNav: function() {
         let tab = `#${opus.prefs.view}`;
+        let browse = (opus.prefs.view === "browse" ? "browse" : "cart_browse");
```
This `let browse =` is duplicated many times. We should have a function for it.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -978,28 +976,33 @@ var o_browse = {
+        let suppressScrollY = false;
+
+        if (opus.prefs[browse] == "gallery") {
+            $(".op-dataTable-view", tab).hide();
```
Shouldn't this be `op-data-table-view` to match `op-data-table`?

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -978,28 +976,33 @@ var o_browse = {
 
     updateBrowseNav: function() {
         let tab = `#${opus.prefs.view}`;
```
I thought we were going to abstract this away so that opus.prefs.view was not used directly as a selector.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -589,6 +596,10 @@ var opus = {
 $(document).ready(function() {
+    // just for browse and cart, as we need to abstract some functions easier
+    // needs to be done in the document.ready because o_cart & o_browse need to be defined first
+    opus.namespaceDict = { "cart": o_cart, "browse": o_browse };
```
If this maps "view" to namespace, I think it should have a name that reflects that. Like viewToNamespaceDict.

dstopp committed:
use correct tab; minor bug fixes
add namespace/tab get funcs to default to browse for loadData purposes

dstopp committed:
Merge branch 'new_ui' of https://github.com/SETI/pds-opus into c035_cart

# Conflicts:
#	opus/application/static_media/js/browse.js

dstopp committed:
-added cartCount to o_cart
-fixed blank cart issue

dstopp committed:
-clean up some class/id names
-fix sort order click issue
-added method for cart_browse

dstopp committed:
cleanup of more class names.

dstopp committed:
missed this file

dstopp committed:
minor change to mutationObserver; class name update

dstopp committed:
this fixes b086 - race condition.  Needed to check for undefined variable; it's not actually a race condition.

dstopp committed:
fixed the incorrect tab on the init of infinite scroll objects

dstopp committed:
calculate and count the gallery images before infiniteScroll object is created so the limit should never be 0

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1558,26 +1563,43 @@ var o_browse = {
 
     countGalleryImages: function() {
         let tab = opus.getViewTab();
```
countGalleryImages needs to take a tab argument like everything else.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1356,14 +1359,16 @@ var o_browse = {
+    initInfiniteScroll: function(view, selector) {
+        let tab = `#${view}`;
+        let startObsLabel = o_browse.getStartObsLabel(view);
+        opus.getViewNamespace(view).galleryBoundingRect = o_browse.countGalleryImages();
```
Note that countGalleryImages uses the "current" opus.prefs.view tab, while the rest of this function uses the view that was passed in. countGalleryImages needs to take a tab argument as well.

rfrenchseti replied:
getLimit() below throughout the path function (many instances) needs to take a tab argument

rfrenchseti replied:
also getViewNamespace needs to take a tab argument

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1298,6 +1298,9 @@ var o_browse = {
 
         // need to add limit - getting twice as much so that the prefetch is done in one get instead of two.
         let limitNum = customizedLimitNum === undefined ? o_browse.getLimit() * 2 : customizedLimitNum;
+        if (limitNum === 0) {
```
Remove debugging code eventually.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1298,6 +1298,9 @@ var o_browse = {
 
```
updateStartobsInUrl is using the current opus.prefs.view. It needs to use the passed-in tab.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1298,6 +1298,9 @@ var o_browse = {
 
         // need to add limit - getting twice as much so that the prefetch is done in one get instead of two.
         let limitNum = customizedLimitNum === undefined ? o_browse.getLimit() * 2 : customizedLimitNum;
```
getLimit() needs to take the tab as an argument

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1558,26 +1563,43 @@ var o_browse = {
+        let width = o_browse.calculateGalleryWidth();
+        let height = o_browse.calculateGalleryHeight();
+
+        let xCount = Math.floor(width/o_browse.imageSize);   // images are 100px
```
Probably want to kill the comment giving a hard value for the image size

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1558,26 +1563,43 @@ var o_browse = {
+        return  $(window).height()-totalNonGalleryHeight;
+    },
+
+    calculateGalleryWidth: function() {
```
calculateGalleryWidth and Height need to take tab arguments

dstopp committed:
continuing to address race condition issues.

dstopp committed:
update the view in the hash instead of using what's in the current hash, for race conditions.

dstopp committed:
just removing some comments and debug code.

dstopp committed:
typo

dstopp committed:
dang, needed a parseInt for the css value or else it returned NaN

dstopp committed:
fix for table delete on tab change; also only update cart on cartChange

dstopp committed:
maxCachedObservations set to 750
add view to loadData for more race condition stuff
mostly whitespace cleanup for long lines

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -672,7 +682,10 @@ var o_browse = {
             // Fot gallery view, the topBoxBoundary is the top of .gallery-contents
             // For table view, we will set the topBoxBoundary to be the bottom of thead (account for height of thead)
-            topBoxBoundary = topBoxBoundary || (opus.prefs[browse]  === "gallery") ? $(`${tab} .gallery-contents`).offset().top : $(`${tab} .gallery-contents`).offset().top + $(`${tab} .op-data-table thead th`).outerHeight();
+            topBoxBoundary = topBoxBoundary ||
```
Add extra level of parens around expression

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -36,7 +36,8 @@ var o_browse = {
     galleryData: {},  // holds gallery column data
     imageSize: 100,     // default
-    maxCachedObservations: 1000,    // max number of obserations to store in cache; at some point, we can probably figure this out dynamically
+    maxCachedObservations: 750,    // max number of obserations to store in cache;
```
obserations => observations

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1033,6 +1051,17 @@ var o_browse = {
     },
 
+    updateViewInUrl: function(view, url) {
+        // remove any existing view= slug or startobs= slug
```
Remove or startobs= slug

dstopp committed:
minor cleanup
